### ISSUE TITLE
TopK operator spec ambiguity

### ISSUE THREAD

**Issue by hariharans29** [onnx/onnx#2075]
Hello all,

The current `TopK` operator is unclear as to whether the value of `k` can be larger than the total number of elements in a particular axis and if so can a backend still proceed with the operation by performing k = min(k, number_of_elements) or if it has to strictly expect that the number of elements be greater than or equal to `k`. 

We propose going with k = min(k, number_of_elements) approach (and clarifying this in the spec) as sometimes the caller doesn't know the exact number of elements in an axis (number of objects detected for example) and `k` itself is usually configured to be a specific value.

Any input will be appreciated. Thanks!

CC: @jiafatom

**Comment by wschin** [comment]
`k = min(k, number_of_elements)` could lead to zig-zag tensors. For example, the first row of a matrix contains 10 elements (`k=10`) and the second row of that matrix has only 2 elements. Would you think use `padding value` a better idea?

**Comment by jiafatom** [comment]
@wschin we have seen some tensorflow models which applies `k = min(k, number_of_elements)` before TopK operation. This needs three additional ops (shape, slice, min), so we are thinking that we want to do it inside TopK.

It is true that the result will be zig-zag. People do padding in this case, but they do padding after TopK, not inside TopK. So for our case, we usually need padding after TopK. Does this address your concern on padding?

**Comment by wschin** [comment]
No. There is no zig-zag tensor in ONNX, so no operator can produce zig-zag tensor.

**Comment by jiafatom** [comment]
@wschin maybe I misunderstand what you mean "zig-zag" tensor. Actually k is a scaler, so I don't quite get why the first row contains 10 elements and second row has 2 elements.

**Comment by wschin** [comment]
Do you have a small TF sample which includes the conversion you need?

**Comment by wschin** [comment]
I probably get your point. So, instead of proposing large operator to support different configurations, I personally prefer smaller primitives. As you said, we can use (`Shape`, `Min`, `topK`) to compose the desired behavior. Note that calling `Shape` and `Min` should be very cheap, so we can ignore the induced cost. Thanks.

**Comment by jiafatom** [comment]
We need two additional Cast op for this. The input of Min should be double/float, so I need cast the shape to float. After that, it needs cast back to integer to feed as TopK input. Then it needs Shape, Min, two cast, which are 4 ops. 

If we do it inside topK, we only need 2-3 lines of code change.

**Comment by wschin** [comment]
Two addition `Cast`s are fine. They all operate on `scalars` so the computation cost can be ignored. It looks like the expressiveness of ONNX is enough right? It makes me wonder the benefit of merging all the 4 operators into one `TopK` if the computation cost is the same?

**Comment by jiafatom** [comment]
I am now working on converting a keras model to onnx. For some custom layers, I need manually convert it - it means that I need to combine onnx node explicitly together. So I need create these 4 ops and link them together for TopK case. If in future we still have TopK, then we should do this again. This is error-prone labor work at converter side. 

But if I change on onnxruntime, it is just 2-3 lines code change, very straightforward and easy to maintain.

**Comment by wschin** [comment]
Sounds like we should improve converter. If you view ONNX as an alternative of C++, you probably don't want to change C++ standard to save 2-3 lines in g++.

**Comment by gramalingam** [comment]
Hi: I don't see why the result will be zig-zag (different number of elements at different positions), since the number of elements is the same in every position of the axis. (If there are repeated values, they should be repeated in the output too.) The original proposal sounds fine to me … the size of the output (along axis) will be the minimum of k and original dimension along axis. We could also provide a padding-value if there are models/use-cases where it is useful to pad upto k.

**Comment by wschin** [comment]
> Hi: I don't see why the result will be zig-zag (different number of elements at different positions), since the number of elements is the same in every position of the axis. (If there are repeated values, they should be repeated in the output too.) The original proposal sounds fine to me … the size of the output (along axis) will be the minimum of k and original dimension along axis. We could also provide a padding-value if there are models/use-cases where it is useful to pad upto k.

It's not zig-zag. I misread the original message. 

My point is that `k = min(k, number_of_elements)` can be done by using a sub-graph with ignorable computation cost. Note that the `k` in `TopK` is an input tensor. Why do we prefer to change the signature of `TopK`?

**Comment by gramalingam** [comment]
The problem is that the spec is incomplete if we do not say what will happen in cases where k > number_of_elements. An incomplete spec can lead to different implementations doing different things. How do we make the spec complete? Either we need to say that (a) the output will have min(k, number_of_elements) or that (b) It is an undefined situation (throw exception?), etc. etc. It seems to me that (a) is a better choice. May be this was what was originally intended? In which case, we are just clarifying things.

**Comment by gramalingam** [comment]
Of course, (c) output will be padded is a third choice.

**Comment by jiafatom** [comment]
@wschin This change does not change signature of `TopK`. 
if `number_of_element >= k`, the behavior does not change.
otherwise, the current behavior will throw exception. This change will change `k = min(k, number_of_elements)` and proceed.

To avoid crash, someone must be responsible to do the `min` operation, either the customer or onnxruntime. Currently, customers must do a min operation to avoid exception. Feels more reasonable that we handle the exception.

**Comment by jiafatom** [comment]
I don't prefer padding inside `TopK`, because padding does change `TopK` behavior. Generally the model owner will add padding after `TopK`.

**Comment by wschin** [comment]
Pytorch and Tensorflow `TopK` throw if there is no enough elements, so I don't feel right to define ONNX `TopK` to incorporate pre-processing steps. Could we explicitly say (b)?

**Comment by hariharans29** [comment]
I prefer (a) as well - do min (k, number_of_elements). 

Right now, in ONNX Runtime we do (b) (i.e.) throw an exception.

@wschin - I did see that TF doesn't allow this, but ONNX doesn't necessarily comply with TF operation as long as the spec is clear and well documented right ?

Thanks!

**Comment by wschin** [comment]
ONNX doesn't need to follow TF and Pytorch for sure. 

Let's consider one situation I am trying to avoid. Let's create a `TopK` operator and save it as an ONNX model called `onnxModel`. When loading `onnxModel` back to pytorch, pytorch's ONNX loader will have to do something to avoid throwing because that ONNX `TopK` won't throw. If we go with (a) or (c), it will incur some necessary changes in ONNX to Pytorch converters. The same to TF case.

**Comment by hariharans29** [comment]
@wschin - I see your concern. Thanks for elaborating.
